File: tools/frame_parser.py
from __future__ import annotations
from dataclasses import dataclass, field

from tools.protocol import MAGIC, FrameType


@dataclass
class ChannelInfo:
    channel_id:   int
    channel_type: int
    name:         str


@dataclass
class DeviceInfoFrame:
    fw_major: int
    fw_minor: int
    channels: list[ChannelInfo] = field(default_factory=list)


@dataclass
class StreamDataFrame:
    channel_id: int
    seq:        int
    samples:    list[int]


@dataclass
class AckFrame:
    cmd_type:    int
    status:      int
    config_data: bytes = b""  # remaining bytes for CONFIG acks


AnyFrame = DeviceInfoFrame | StreamDataFrame | AckFrame


class FrameParser:
    def __init__(self) -> None:
        self._buf:      bytearray      = bytearray()
        self._last_seq: dict[int, int] = {}
        self.seq_drops: int            = 0

    def feed(self, data: bytes | bytearray) -> list[AnyFrame]:
        self._buf += data
        frames: list[AnyFrame] = []
        while (f := self._try_parse()) is not None:
            frames.append(f)
        return frames

    def _try_parse(self) -> AnyFrame | None:
        while True:
            # Resync: drop bytes until 0xAD or buffer empty
            while self._buf and self._buf[0] != MAGIC:
                del self._buf[0]

            if len(self._buf) < 4:
                return None

            type_byte   = self._buf[1]
            payload_len = (self._buf[2] << 8) | self._buf[3]
            frame_size  = 4 + payload_len

            if len(self._buf) < frame_size:
                return None

            payload = bytes(self._buf[4:frame_size])
            del self._buf[:frame_size]

            result = self._decode(type_byte, payload)
            if result is not None:
                return result
            # Unknown type: payload already consumed by LENGTH, loop for the next frame.

    def _decode(self, type_byte: int, payload: bytes) -> AnyFrame | None:
        match type_byte:
            case FrameType.DEVICE_INFO:
                return self._decode_device_info(payload)
            case FrameType.STREAM_DATA:
                return self._decode_stream_data(payload)
            case FrameType.ACK:
                return self._decode_ack(payload)
            case _:
                return None

    def _decode_device_info(self, payload: bytes) -> DeviceInfoFrame | None:
        if len(payload) < 3:
            return None
        fw_major, fw_minor, ch_count = payload[0], payload[1], payload[2]
        channels: list[ChannelInfo] = []
        i = 3
        for _ in range(ch_count):
            if i + 3 > len(payload):
                break
            ch_id, ch_type, name_len = payload[i], payload[i + 1], payload[i + 2]
            i += 3
            name = payload[i:i + name_len].decode("ascii", errors="replace")
            i += name_len
            channels.append(ChannelInfo(channel_id=ch_id, channel_type=ch_type, name=name))
        return DeviceInfoFrame(fw_major=fw_major, fw_minor=fw_minor, channels=channels)

    def _decode_stream_data(self, payload: bytes) -> StreamDataFrame | None:
        if len(payload) < 5:
            return None
        ch_id = payload[0]
        seq   = (payload[1] << 8) | payload[2]
        count = (payload[3] << 8) | payload[4]
        if len(payload) != 5 + count * 2:
            return None
        samples = [
            (payload[5 + i * 2] << 8) | payload[5 + i * 2 + 1]
            for i in range(count)
        ]
        last = self._last_seq.get(ch_id)
        if last is not None:
            expected = (last + 1) & 0xFFFF
            if seq != expected:
                self.seq_drops += (seq - expected) & 0xFFFF
        self._last_seq[ch_id] = seq
        return StreamDataFrame(channel_id=ch_id, seq=seq, samples=samples)

    def _decode_ack(self, payload: bytes) -> AckFrame | None:
        if len(payload) < 2:
            return None
        return AckFrame(cmd_type=payload[0], status=payload[1],
                        config_data=payload[2:])

File: host/src/frame_parser.cpp
#include "frame_parser.hpp"

std::optional<FrameParser::Frame> FrameParser::try_parse()
{
    while (true) {
        // Resync: drop bytes until buf_[0] == MAGIC or buffer empty
        while (!buf_.empty() && buf_[0] != MAGIC)
            buf_.erase(buf_.begin());

        // Need at least header: magic(1) + type(1) + length(2)
        if (buf_.size() < 4)
            return std::nullopt;

        uint8_t  type        = buf_[1];
        uint16_t payload_len = (uint16_t(buf_[2]) << 8) | buf_[3];
        size_t   frame_size  = 4 + payload_len;

        if (buf_.size() < frame_size)
            return std::nullopt;

        if (type != STREAM_DATA || payload_len < 5) {
            // Unknown type or malformed STREAM_DATA: skip via LENGTH, loop for next
            buf_.erase(buf_.begin(), buf_.begin() + frame_size);
            continue;
        }

        // STREAM_DATA payload: ch_id(1) seq(2BE) count(2BE) samples[](uint16BE)
        uint8_t  ch_id = buf_[4];
        uint16_t seq   = (uint16_t(buf_[5]) << 8) | buf_[6];
        uint16_t count = (uint16_t(buf_[7]) << 8) | buf_[8];

        if (payload_len != uint16_t(5 + count * 2u)) {
            buf_.erase(buf_.begin(), buf_.begin() + frame_size);
            continue;
        }

        Frame f;
        f.channel_id = ch_id;
        f.seq        = seq;
        f.samples.resize(count);
        for (size_t i = 0; i < count; ++i)
            f.samples[i] = (uint16_t(buf_[9 + i * 2]) << 8) | buf_[9 + i * 2 + 1];

        if (last_seq_) {
            uint16_t expected = uint16_t(*last_seq_ + 1);
            if (seq != expected)
                seq_drops_ += uint16_t(seq - expected);
        }
        last_seq_ = seq;

        buf_.erase(buf_.begin(), buf_.begin() + frame_size);
        return f;
    }
}

std::vector<FrameParser::Frame> FrameParser::feed(std::span<const uint8_t> data)
{
    buf_.insert(buf_.end(), data.begin(), data.end());

    std::vector<Frame> frames;
    while (auto f = try_parse())
        frames.push_back(std::move(*f));
    return frames;
}
